### ISSUE TITLE
How to style specific nested tokens?

### ISSUE THREAD

**Issue by adamwathan** [react-syntax-highlighter/react-syntax-highlighter#204]
**Is your feature request related to a problem? Please describe.**

The built-in styles include nested token styles like this:

https://github.com/conorhastings/react-syntax-highlighter/blob/master/src/styles/prism/ghcolors.js#L163

However as far as I can tell they don't actually seem to work.

**Describe the solution you'd like**

It would be great if there was some way to target nested tokens like you can when writing a CSS theme for Prism/Highlight.

**Describe alternatives you've considered**

Tried a bunch of syntaxes to see if anything worked:

```
<SyntaxHighlighter
  style={{
    '.token.atrule .token.rule': { background: 'pink' }
    'atrule rule': { background: 'pink' }
    'atrule.rule': { background: 'pink' }
  }}
  {/* ... */}
/>
```

**Comment by karlhorky** [comment]
Just ran into this today, trying to style `interpolation-punctuation`.

Simple reproduction here:

![Screen Shot 2020-08-24 at 14 21 28](https://user-images.githubusercontent.com/1935696/91044355-70d60f80-e615-11ea-83dc-56454bd5dc9a.png)

Sandbox: https://codesandbox.io/s/react-syntax-highlighter-forked-kjnk7?file=/src/index.js:265-559

**Comment by karlhorky** [comment]
@simmerer I would like to try my hand at doing a PR for this.

I've narrowed it down to this part of the `create-element.js` file:

https://github.com/react-syntax-highlighter/react-syntax-highlighter/blob/8aacf945df624a9e32143f6bea74c0a6f8a92825/src/create-element.js#L3-L7

For example, if the `classNames` variable looks like this, which it does for the case of `interpolation-punctuation`:

```js
["token", "template-string", "token", "interpolation", "token", "interpolation-punctuation", "punctuation"]
```

I can think of two ways to solve this:

1. Simpler, naïve solution: make the order of the `classNames` array go from less specific to more specific (in terms of Prism.js hierarchy)
2. More involved, potentially more "correct" solution: sort the keys of the stylesheet by CSS specificity and iterate over them instead of the `classNames`

**Comment by karlhorky** [comment]
I will come up with a prototype of the specificity solution above (solution 2).

**Comment by karlhorky** [comment]
Ok so using [`specificity`](https://www.npmjs.com/package/specificity), we can do this solution (working, but not yet optimized):

```diff
import React from 'react';
+import { compare } from 'specificity';

export function createStyleObject(classNames, elementStyle = {}, stylesheet) {
+  const sortedStylesheet = Object.entries(stylesheet).sort(([selectorA], [selectorB]) => {
+    return compare(selectorA, selectorB)
+  });
-  return classNames.reduce((styleObject, className) => {
-    return { ...styleObject, ...stylesheet[className] };
+  return sortedStylesheet.reduce((styleObject, [selector, styles]) => {
+    return { ...styleObject, ...(classNames.includes(selector) && styles) };
  }, elementStyle);
}
```

**Comment by karlhorky** [comment]
Hmm... I guess this doesn't help actually, since `classNames` is this:

```js
["token", "template-string", "token", "interpolation", "token", "interpolation-punctuation", "punctuation"]
```

So the `styles` are like this (same specificity):

```js
{
  punctuation: {
    color: '#ff0000',
  },
  'interpolation-punctuation': {
    color: '#00ff00',
  },
}
```

**Comment by karlhorky** [comment]
Ok, found a zero-dependency alternative, which:

1. Splits up the selectors in the `stylesheet` by `.`s into an array
2. Checks whether this array is a subset of the `classNames` array
3. Adds styles related to class selectors to the end of the styles to be more specific

Feedback on this approach?

```diff
+function classSelectorToClassNames(selector) {
+  if (!selector.includes(".")) return null;
+  return selector.split(".").filter(Boolean);
+}
+
+// Is array1 a subset of array2
+function isSubset(array1, array2) {
+  if (!array1 || !array2) return false;
+  if (!Array.isArray(array1) || !Array.isArray(array2)) return false;
+  const intersection = array1.filter((element) => array2.includes(element));
+  return intersection.length === array1.length;
+}

export function createStyleObject(classNames, elementStyle = {}, stylesheet) {
-  return classNames.reduce((styleObject, className) => {
-    return { ...styleObject, ...stylesheet[className] };
-  }, elementStyle);
+  const newStyles = Object.entries(stylesheet).reduce(
+    (newStyleObject, [selector, styles]) => {
+      if (classNames.includes(selector)) {
+        return { ...styles, ...newStyleObject };
+      } else if (isSubset(classSelectorToClassNames(selector), classNames)) {
+        return { ...newStyleObject, ...styles };
+      }
+      return newStyleObject;
+    },
+    {}
+  );
+  return { ...elementStyle, ...newStyles };
}
```

**Comment by karlhorky** [comment]
Opened a PR with this approach at https://github.com/react-syntax-highlighter/react-syntax-highlighter/pull/305 - of course open for feedback and changing the approach!

**Comment by karlhorky** [comment]
@adamwathan a solution from #305 has now been merged and is available in `react-syntax-highlighter@15.0.1`! 

If it works for you, maybe this issue can be closed :)
